### PR TITLE
Unify hasConvertToMMATransisitiveUse

### PR DESCRIPTION
**Pull request by htyu**
Unifying some code in RemoveLayoutConversions.cpp between NV and ROCm. The current ROCm path misses some update from the main repo.

### CODE REVIEW

File: lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp [hunk 1]
```diff
@@ -171,47 +171,21 @@ static bool hasConvertToMMATransisitiveUse(Operation *op, Attribute encoding) {
...
-        continue;
-      auto forOp = dyn_cast<scf::ForOp>(yield.getOperation()->getParentOp());
-      if (!forOp)
-        continue;
-      for (OpOperand &operand : yield->getOpOperands()) {
-        Operation *def = operand.get().getDefiningOp();
-        if (def && forwardSlice.count(def) &&
-            (seen.insert(operand.get()).second == true))
-          queue.push_back(forOp.getRegionIterArg(operand.getOperandNumber()));
+        if (auto mfmaLayout =
```
Comment (scxiao) [inline lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp:174]: Can we rename the function name `hasConvertToMMATransisitiveUse` to be general?

**Comment by zhanglx13** [comment]
@htyu It's better to put the unifying between nv and amd kind of the work in upstream. And we've not started such work because the refactor in upstream is not settled yet. 
Is this PR related to correctness or performance? If not, I'd suggest we do it later at upstream afterwards.

**Comment by htyu** [comment]
> @htyu It's better to put the unifying between nv and amd kind of the work in upstream. And we've not started such work because the refactor in upstream is not settled yet. Is this PR related to correctness or performance? If not, I'd suggest we do it later at upstream afterwards.


Sounds good to move this to upstream. It's for performance. The RocM path is missing some recent work on the NV path.

**Comment by htyu** [comment]
> And we've not started such work because the refactor in upstream is not settled yet.

I see. It might sounds better to pull the NV changes onto RocM path instead of refactoring for unification?

**Comment by zhanglx13** [comment]
> It might sound better to pull the NV changes onto RocM path instead of refactoring for unification?

Upstream has done a lot of refactor in the frontend, which makes the "pull the NV changes" process very painful and unnecessary for performance.
If you need this PR to bring up the perf of your kernel, we can merge it for now. And worry about the big upstream unification later.

File: lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp (hunk 1 above)
Comment (htyu) [reply lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp:174]: Both `MfmaEncodingAttr` and `NvidiaMmaEncodingAttr` have `MmaEncodingTrait`, so `MMA` here sounds genernal?

**Comment by htyu** [comment]
> > It might sound better to pull the NV changes onto RocM path instead of refactoring for unification?
> 
> Upstream has done a lot of refactor in the frontend, which makes the "pull the NV changes" process very painful and unnecessary for performance. If you need this PR to bring up the perf of your kernel, we can merge it for now. And worry about the big upstream unification later.

Thanks for your understanding! Sounds good to land it here.

**Comment by htyu** [comment]
BTW, do you think the test failure (fused attention) is related? The archived log is empty. I was also trying to repro it locally by installing the nightly pytorch

`pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/rocm5.6/
`
but then I hit 
`TypeError: _attn_fwd() got an unexpected keyword argument 'slice_k_tile'`

Is it the right way to get a repro?

**Comment by zhanglx13** [comment]
@htyu The pytorch wheel comes with a pre-installed version of triton, which is pretty old.
You'll need to install triton from this repo.

**Comment by zhanglx13** [comment]
I tried it locally and got the following error 
`LLVM ERROR: Unimplemented usage of getCTALayout`

**Comment by htyu** [comment]
> I learned that pytorch 2.3 has some issues with rocm5.6. You can use the original pytorch comes with the rocm5.6 docker and re-install triton.

I see. Docker isn't available on our server unfortunately :( For now I just fixed the error as you pointed out. Let's see how it works.

**Review by zhanglx13 — APPROVED** [review]
Approved since it matters to your kernel @htyu.
But I'm just wondering why you need to add `getCTALayout` for mfma. It usually means you are adding conversion from mfma to other layouts that requires LDS to shuffle the data.

**Pull request merged**